src/zope/lifecycleevent/README.rst

```diff
@@ -0,0 +1,196 @@
+:class:`~IObjectAddedEvent` and its implementation
+:class:`~.ObjectAddedEvent`, or the API
+:func:`zope.lifecycleevent.added`.
+
+    >>> from zope.lifecycleevent import ObjectAddedEvent
+    >>> from zope.lifecycleevent import added
+
+    >>> container = {}
+    >>> container['name'] = obj
+    >>> added(obj, container, 'name')
+
+
+    >>> obj['modified'] = True
+
+
+Modification
+============
+
+One of the most common types of events used from this package is the
+:class:`IObjectModifiedEvent` (implemented by
+:class:`ObjectModifiedEvent`) that represents object modification.
+
+In the simplest case, it may be enough to simply notify interested
+parties that the object has changed. Like the other events, this can
+be done manually or through the convenience API
+(:func:`zope.lifecycleevent.modified`):
+
+    >>> obj['key'] = 42
+
+    >>> from zope.lifecycleevent import ObjectModifiedEvent
+    >>> notify(ObjectModifiedEvent(obj))
+
+The above is equivalent to this:
+
+    >>> from zope.lifecycleevent import modified
+    >>> modified(obj)
+
+Providing Additional Information
+--------------------------------
+
+Some event consumers like indexes (catalogs) and caches may need more
+information to update themselves in an efficient manner. The necessary
+information can be provided as optional "modification descriptions" of
+the :class:`~.ObjectModifiedEvent` (or again, via the
+:func:`.zope.lifecycleevent.moved` function).
+
+This package doesn't strictly define what a "modification description"
+must be. The most common (and thus most interoperable) descriptions
+are based on interfaces.
+
+We could simply pass an interface itself to say "something about the
+way this object implements the interface changed":
+
+    >>> from zope.interface import Interface, Attribute, implementer
+    >>> class IFile(Interface):
+    ...     data = Attribute("Data")
+    >>> @implementer(IFile)
+    ... class File(object):
+    ...     pass
+
+    >>> file = File()
+    >>> file.data = "123"
+    >>> notify(ObjectModifiedEvent(obj, IFile))
+
+We can also be more specific in a case like this where we know exactly
+what attribute of the interface we modified. There is a helper class
+:class:`zope.lifecycleevent.Attributes` that assists:
+
+    >>> file.data = "abc"
+    >>> from zope.lifecycleevent import Attributes
+    >>> modified(obj, Attributes(IFile, "data"))
+
+When an interface attribute is a sequence or container, we can specify
+the individual keys that we changed using
+:class:`zope.lifecycleevent.Sequence`:
+
+    >>> class IFiles(Interface):
+    ...    files = Attribute("Sequence of IFile objects")
+    >>> @implementer(IFiles)
+    ... class Files(object):
+    ...    pass
+
+    >>> files = Files()
+    >>> files.files = [File()]
+    >>> created(files)
+
+    >>> from zope.lifecycleevent import Sequence
+    >>> files.files.append(File())
+    >>> modified(files, Sequence(IFiles, len(files.files) - 1))
```

Comment:
My reading of the Sequence docstring leads me to believe it's more for cases like

```python
class IFileList(ISequence):
    pass

@implementer(IFileList)
class FileList(list):
    pass

files = FileList()
files.append(File())
modified(files, Sequence(IFileList, len(files) - 1))
```